### DISCUSSION TITLE
Identifing websocket client session when using httpOnly cookie and auth callout

### DISCUSSION THREAD

**Discussion by jonaskello** [nats-io/nats-server#5143] [General]
Now that we have the possibility to use [httpOnly cookies to pass a value to auth callout](https://github.com/nats-io/nats-server/pull/5131) it is almost possible to only use httpOnly cookie when doing NATS rpc call from the browser to replace regular HTTP request/response. The final hurdle I have is to be able to identity the calling client's session without leaking that information to the javascript in the browser (for regular HTTP request this is fully solved using httpOnly cookie).

What we are doing is setting a httpOnly cookie with a `sessionId` in a traditional way from a regular HTTP backend. This `sessionId` is stored in a backend memory database (redis) together with other information such as tokens needed to make other backend calls etc. Then we connect with `nats.ws` from the browser to NATS server and since we have `token_cookie` configured this `sessionId` is forwarded to the auth callout service. The auth callout service validates the session on the backend and issues a JWT. All this can be done without leaking the sessionId to the browser code since we can use httpOnly cookie for this whole flow.

However we also need to be able to know which sessionId is making RPC calls into NATS. For example a service running in NATS may have to make calls using one of the tokens associated with the session, so it needs to get session info from redis. This is possible in several ways already for example we are currently putting the sessionId in the subject. However this is no longer possible with httpOnly cookie since we cannot read the cookie in the browser code. So we need some other way to do this part if we want to use httpOnly cookie (which we want because it is more secure). I have considered the following solutions:

1. Putting the `sessionId` in the subject and limit the issued JWT to something like `{sessionId}.foo.getBar`. This works and allows the service receiving the call to have the sessionId similar to how it gets it in a cookie header for a regular HTTP request. However as I mentioned above it requires the browser javascript to have the `sessionId` available in order to contruct the subject. This means we cannot have it only as httpOnly cookie and thus it is leaked into the client code.

2. Getting hold of the NATS `client_id` field in the auth callout and restrict the subject to `{client_id}.foo.getBar`. Then somehow get hold of the `client_id` in the browser (I don't think `nats.ws` currently exposes this but from what I understand it is returned from `INFO` call in the NATS protocol). On the backend in the auth callout server associate the `client_id` with the `sessionId`. This avoids exposing the `sessionId` in the client code but in RPC calls we can extract the `client_id` from the subject and then use it to lookup the `sessionId`. As an alternative we could generate some unique ID in the auth callout servcie and use that instead of `client_id`. But I don't think there is a way to communicate something like this back to the connecting client from the auth callout service?

3. In the JWT that the auth callout issues put subject mapping rules so that `foo.getBar` is mapped to `the_session_id.foo.getBar`. This way we do not need to expose the `sessionId` to the browser and NATS will automatically add it to the subjects by mapping. However from what I understand the JWT that is issued by auth callout cannot have subject mappings, only account tokens can have this.

Guidance and more ideas are very welcome!

**Comment by aricart** [comment]
This has nothing to do with the connection or the session per se. NATS interactions are anonymous. If you have permission, you can, but who is making the request or responding is only known if you include additional information. Now, with that said, the correct pattern for this is to encode the id/session information as part of the subjects. The client would only be able to publish messages including this ID (the server that set the auth cookies could reveal this to the web application so that the client can setup itself with an inbox that looks like `_INBOX.aid.someid.>` as a prefix, and then for the request subjects publish to `q.someservice.aid.someid`, the callout should ensure that the client only has permissions to subscribe to things that include such pattern correctly, ensuring that no other client can publish or subscribe to messages intended for the particular client. Your application then, knowing this session can figure out who the client is and behave accordingly. There's a possibly outdated example that you can look at for some insight that used this strategy (https://github.com/aricart/authdemo).

**Reply by jonaskello** [reply]
> the correct pattern for this is to encode the id/session information as part of the subjects

Yes we are currently doing exactly this and technically this works. However what we found is that when you use this pattern in the web browser it becomes vulnerable to cross site scripting (XSS) attacks. I realise now that I should have included this information in the initial post, sorry about that.

The short story about XSS attacks is that any value that is exposed to javascript in the browser can also be stolen by malicious actors. The javascript in the browser using the NATS client needs to know `someid` (that is the session id) in order to publish to a subject that contains it. By making `someid` available to the javascript running the NATS client in the browser, it is also made available for theft by malicious actors.

**Reply by jonaskello** [reply]
In HTTP you can fully block XSS attacks by using a httpOnly cookie. The way it works is that the server after authentication sets a httpOnly cookie with a session identifier. The browser stores this cookie but does not make it available to the javascript running in the browser. So there is no way for malicious actors to steal this cookie value by injecting javascript. The browser will pass this cookie to the server on each request so the server can identify the session.  So with httpOnly cookies, the server can set a cookie value that it gets back on every request but no javascript in the browser can see this cookie value, making it fully resistent to XSS attacks.

What we are looking for is something similar for NATS rpc calls from the browser.

**Reply by jonaskello** [reply]
I made a proposal for one way of solving it in #5147

**Reply by aricart** [reply]
It won't be - the permissions that you are minting will not allow any other client to publish on this - so it is inconsequential if the id is seen (obviously make it a hash) - the auth token is never seen by the client directly so access is secure.

**Reply by aricart** [reply]
Also, remember that you can always make your HTTP server proxy the interactions with NATS; in that case, you can simplify a lot of the interactions and handle all the security within your app rather than have the browser reach in.

**Reply by jonaskello** [reply]
> the permissions that you are minting will not allow any other client to publish on this

Yes this is 100% true and we are already doing it this way. By doing this it is 100% certain that the client is the only one being able to publish because the connection with the corresponding token is the only one that has rights to publish on the subject. We are fully aware of this part :-). Looking only on this part in isolation it is secure. This is not the problem.

>  so it is inconsequential if the id is seen

It is inconsequential in the way that it just has to match the one in the token to make sure that the client is the only one that can publish on it. But it is not inconsequential in the way that the id itself must be OK to may be stolen by an XSS attack without causing harm. You may think that even if someone stole the id then it is still no big deal since they can not publish on the corresponding subject since they do not have the connection with the corresponding token that allows publishing with that id. In theory this is correct if you just look at this particular part of the flow. However if you consider the full flow trying to make this secure inside the browser is in practice not really possible.

The problem is that in order for the id to be useful it must be meaningful to the server. So if the client just invents some id that it sends along when connecting that would not be meaningful to the server. So the server must be the one to create the id and it must associate it with something on the server-side like a user session. So the problem lies with the server having to create the id and associate it with something and that XSS attacks are possible. Maybe it is easier to explain with one of the full flows that does not work as an example:

1. The HTTP server logs the user in and sets a httpOnly cookie with the sessionid.
2. The browser websocket nats client connects sending along the sessionid from the cookie.
3. The auth callout service is called and receives the sessionid.
4. The auth callout service creates a new unique id to be used in the subject (rather than using the sessionid becuase that should not be exposed to the browser). It adds this id to the minted JWT subjects so only this client can send on subjects that have this unique id. It also associates this new unique id with the sessionid so when services later gets the id on the subject they can lookup the session id.
6. Here it breaks down because there is no way (that I know of) for the auth callout service to communicate the generated unique id back to the browser client that is connecting. So it will have no way of knowing that it should use this id in the subjects it publishes on. Even though it would be OK to have that id exposed in the browser and stolen by XSS attack as it is not the session id and it will always be re-generated so it cannot be used for new connections. If there was some mechanism to communicate the id back to the client then this flow could work.

> Also, remember that you can always make your HTTP server proxy the interactions with NATS; in that case, you can simplify a lot of the interactions and handle all the security within your app rather than have the browser reach in.

The subject is the only way to identify the client. There is no server proxy I know of that can alter NATS subjects because that would mean it has to be able to decode the NATS protocol? So there is no proxy that can help with this AFAIK.

**Reply by aricart** [reply]
What I mean is your application (on the HTTP server) can make the NATS requests for you. Currently you are putting the logic on the browser, but the same could be an URL that you hit that will get all your auth information with each request and you can validate and then do what is needy.

**Reply by jonaskello** [reply]
You mean like writing a custom HTTP gateway and hiding NATS behind it rather than exposing NATS via websockets to the browser? Yes we've considered this as it might be the only way to make things secure.

However using `nats.ws` from the browser has been working very well for us during development and beta. Now that we are moving towards production we need to be more security aware. NATS is also very performant and easy to use for streaming. Introducing a HTTP proxy will introduce extra latency and on top of that streaming does not work well with HTTP so even if we could proxy the RPC calls easily, streams will be more difficult. Simply calling into NATS from the browser is much simpler and more efficient. Only thing is that we cannot find a fully secure flow for it but it is very close. I think #5147 could solve the final piece. Do you have any thoughts about that proposal?

**Reply by aricart** [reply]
The main issue here is that you are asking for the NATS server to do message source attribution, which I am certain breaks an important tenant for NATS.

**Comment by derekcollison** [comment]
If a message is received via a service import information about the client
is provided. Minimal unless the import specified share as true and then
full fidelity of all information about the client is provided in a header.

**Reply by jonaskello** [reply]
Yes I read about this in #4920 that NATS add the `"Nats-Request-Info"` header in this case. We are currently not using imports but perhaps we can re-arrange things to use imports in order to trigger this. I'll try some experiment and see if we can use this.

Btw #4920 and the `"Nats-Request-Info"` header was also what inspired the proposal in #5147 which I think is similar but instead of triggering headers being added by import, they are triggered by the content of the JWT issued for the client.

**Comment by jonaskello** [comment]
After some more experimentation I think I may have a full flow that is secure. It uses the `client_id` that is generated by NATS for each new connection as this value is available both to the auth callout service and the client (I found that it is actually a property of the connection which I did not know before). I think there is guarantee that `client_id` is always unique for each connection and cannot be re-used for new connections. If this is true, then it does not matter if `client_id` is stolen in a XSS attack as the attacker cannot use it for a new connection (which he could with a generic `session_id`).

So the full flow would be:

1. The HTTP server logs the user in and sets a httpOnly cookie with the `session_id` which is stored in the backend database together with other items, such as access tokens needed for making further backend calls. Making the cookie httpOnly means the `session_id` cannot be stolen with XSS.
2. The browser websocket nats client connects sending along the `session_id` from the cookie (using the `token_cookie` websocket config). 
3. The auth callout service is called and receives the `session_id` which it validates with the backend database.
4. The auth callout service finds the  `server_id:client_id`  that the NATS server generated for the connection in the auth request. It puts a mapping between the `server_id:client_id` and the `session_id` in the backend database.
5. The auth callout service issues a token that allows the client to only publish on subject `{server_id:client_id}.>`. This way we can be sure that any messages published on this subject originated from this client.
6. The websocket nats client finds the `server_id:client_id` from the connection properties and prefixes the subject for rpc calls with this, for example `{server_id:client_id}.myService.getFoo`
7. The service that responds to `*.myService.getFoo` extracts the first part of the subject to get the `server_id:client_id` and then looks up the corresponding `session_id` in the backend database. It can then look up any other information associacted with the  `session_id` such as an access token it needs to make a http call to some other service.

It is also possible to use export/import and the `"Nats-Request-Info"` header to replace the need for subject prefixing. This header also contains the `server_id:client_id` IMO this is a nicer solution as it does not pollute the subjects and also simplifies the auth callout service as it des not need to put the prefixed restriction on the subject in the issued token.

**Reply by aricart** [reply]
That is actually a very clever way for identifying the session.

**Reply by ripienaar** [reply]
Client ID is unique per 1 single server.  You can have duplicates on other servers.

**Reply by jonaskello** [reply]
@ripienaar Yes, that was noted below already. We need `server_id:client_id` to make it unique. I updated the flow above to avoid confusion.

**Reply by jhalag** [reply]
While a good workaround, this seems like quite a few steps to go through. I've found several opened issues here with developers looking to find this information - I can't understand why? The `Nats-Request-Info` header looks like it could be useful in _all_ cases - why is it only exposed if imports are used?

I was looking to switch to nats for my service mesh - but this is holding me back. I'd be adding considerable complexity for other developers on my team just for the simple luxury of knowing "who" published a message. It's great that I can trust implicitly that whoever sent the "delete x" call to my microservice was authorized to do so. But the task of logging it for example - I can log that "somebody" deleted the entry via the nats mesh - but not who? Not without a multi-step, multi-tier workaround, or needing to add the configuration overhead of exports/imports?

I suspect I'm missing something here. I was hoping not to use RabbitMQ - but to see multiple closed issues here that state that a service provider should not know the sender of a message has me stumped. RabbitMQ makes it plainly available: https://www.rabbitmq.com/docs/validated-user-id

Am I missing something? For example:

1. If I want it logged that "user X performed action Y" - how is one supposed to do that - _with minimal friction in implementation and configuration_ in nats?
2. Likewise if I have a single microservice that could handle many nuanced requests, and it will need to verify that a given user can perform an action, using service-specific, _user specific_ logic? Splitting out 15 actions to individual topics for individual permission assignment is not palatable.

**Reply by derekcollison** [reply]
Generating the info per message can be penal when you have some message rates that are in the millions. So we limited it to service export/import, but please note that you can do this seamlessly inside the same account, with no impact to the end user applications code. Just export and import to the same account.

**Reply by jhalag** [reply]
That makes a lot of sense. I can see why one would not want those headers appended to high volume, low size messages.

I was not aware that an import/export within the scope of a single account would be valid. In the back of my mind I thought I had read that such would be ignored. That's actually not too heavy for configuration overhead - I could set up wildcard topic(s) for the API entries I need tracking on - and leave the others lean.

Thank you for the clarification @derekcollison !

**Reply by derekcollison** [reply]
No problem. Let us know if you have any issues.

**Reply by jhalag** [reply]
Ah, I found the reference that had me thinking an account could not import from itself: https://docs.nats.io/running-a-nats-service/configuration/securing_nats/accounts#import-configuration-map

There it states _"Accounts cannot do self-imports."_

**Reply by derekcollison** [reply]
Yes that used to be the case, we will update the docs. Thanks for the reminder. /cc @bruth

**Comment by derekcollison** [comment]
They are used only once per server, so it has to be serverid:CID

**Reply by jonaskello** [reply]
Aha, thanks, that's a good catch, I'll update the flow accordingly.

**Reply by jonaskello** [reply]
@derekcollison I noticed that the value for the `name` field that I return from auth callout service is available in the `"Nats-Request-Info"` header as the `user` field. I was thinking I could use this to transfer the `session_id` directly without having to map via `serverid:CID`.

Would it be OK to use the `name` field of the auth callout issued token to store something that the client should not have access to and I just want send back in `"Nats-Request-Info".user` when a service is called?

**Reply by derekcollison** [reply]
You could yes.

**Reply by aricart** [reply]
The client can make a request to `$SYS.REQ.USER.INFO` and get some information back.
Currently the `name` is not telegraphed there - but just want to make you aware that the client can make a request and get some info back, and what other information may be deemed public or added at some point in the future. I would instead rely on the exposed information to obtain information that only the system should have.

**Reply by jonaskello** [reply]
Aha, thanks, I did not know that. You're correct that if the `name` field containing the `session_id` gets exposed to the client then it will be comprimised becuase is not a [nonce](https://en.wikipedia.org/wiki/Cryptographic_nonce) like the `server_id:client_id` field.

The reason for using `name` was that it simplifies the implementation of the auth callout service since in that case it does not need to have access to the session database, and also the services do not need to do double lookup in the session database.

It would be nice if there was a field that auth callout could set in the issued jwt that was guaranteed to not get leaked back to the client, and also was included in `"Nats-Request-Info"`.

**Reply by aricart** [reply]
But `name` is intended as a way of identifying a client on monitoring...

**Reply by jonaskello** [reply]
Yes, as I said above it would be nice if there was another field we could use besides `name` that would be more dedicated to the application's needs. To clarify, the needs of the field are:

1. Possible to set from auth callout.
1. Guaranteed to not get leaked back to the client.
2. Included in `"Nats-Request-Info"`.

I don't think there is such a field already (`name` was the only candidate I found). Just suggesting it would be nice if id could be added.

**Reply by aricart** [reply]
I think that since we only expose fields that are part of the auth, adding additional things without some sort of design contract is not really feasible. Given that you should just use the info there and then retrieve from your back end, it is best for your application to determine what those requirements are rather than try to build something based on some place holders that we provide which won't meet other applications needs.

**Reply by jonaskello** [reply]
From my perspective as an application developer using NATS I think the auth callout feature and the `"Nats-Request-Info"` feature are two huge steps forward to enabling more flexible auth scenarios for applications using NATS. An extra field in the jwt for application specific use would enable any application to add any application specific value and then get it back on each call. I still think that would be yet another good step towards a more flexible auth story. I'm not sure what "design contract" refers to here but IMO the 3 requirements I stated above would be a design contract for this field.
